### DISCUSSION TITLE
Server running in Codespaces occasionally fails to establish forwarded port

### DISCUSSION THREAD

**Discussion by valhuber** [community/community#32788] [Codespaces]
I have a web app project that works 80% of the time - create from scratch like this: https://github.com/ApiLogicServer/ApiLogicProject

But sometimes I see failures, either

- rogue printout in the terminal window
- or, the Browser window fails to initialize (cannot connect to server)

It has always been fine to stop and restart the server.

**Comment by dmilburn** [comment]
Hi @valhuber, thanks for reporting. could I ask you to provide some more information please?

It sounds like you have three distinct issues:

1. the terminal printing out values you don't expect
2. port forwarding doesn't always succeed
3. server doesn't successfully get established

My questions for you:
1. did i accurately understand your problems? or are points 2 and 3 the same?
2. could you provide any screen shots of any of the behavior you're seeing?
3. could you provide any error codes or messages that you're seeing?
4. could you share the name of any codespaces where this has happened?

thank you!

**Reply by valhuber** [reply]
Thanks @dmilburn 

I'd say just one issue, with 2 manifestations...  Port Forwarding fails, either...

1. obviously, in the terminal window (it prints garbage, not just unexpected values), or
2. the server starts cleanly, but the Browser fails to load the page.

My original issue shows the project.  But, be careful, this is a project that _creates_ projects, as explained in the readme link above.  I doubt that's the source of the bug, but it may introduce some confusion if you try to test it.

The screen shot is simply a Browser saying "unable to load page".  I will attach a screen shot next occurrence, but as noted above, it only happens maybe 1 out of 5 times.  I'd suggest you enlist a qa engineer to bang on it...

The project is the one at the original link - follow the readme instructions to load it, and create the project.  Underneath, it's a Python / Flask project.

I am happy to jump on meeting to help clarify.  I think this is an important issue for us both.

**Reply by dmilburn** [reply]
Thanks so much for walking me through the issue and your offer to jump on a call. I'm going to spend some time trying to understand what's happening internally but will definitely reach out if we get blocked. Will circle back soon. Thank you!

**Reply by valhuber** [reply]
@dmilburn -- sounds good.

As I mentioned to Craig, you might want to consider this as a Codespaces demo.  I think it paints quite an impressive picture of Codespaces, VSCode. 

 In fact, I wrote this [article about](https://dzone.com/articles/instant-apps-customize-in-codespaces) it, in addition to [ApiLogicServer/ApiLogicServer readme](https://github.com/ApiLogicServer/ApiLogicProject#readme).

**Reply by dmilburn** [reply]
@valhuber thanks for being such a great Codespaces fan/user! 

1. _the server starts cleanly, but the Browser fails to load the page._

I believe I was able to replicate this issue!

Setting up ApiLogicServer in a codespace, occasionally clicking the globe icon would, instead of opening a url like `*.githubpreview.dev`,  it'd open one starting in `localhost`. Could you confirm whether or not this is what is in the url the next time that "the Browser window fails to initialize"?
 
![image](https://user-images.githubusercontent.com/10053402/190010612-7150801d-4add-4c68-b257-666faaa4f783.png)

2. r_ogue printout in the terminal window_

I haven't been able to replicate this one. Is there any other information you could provide? Is this one equally as random on server start or does it happen after the server has been on for a while? Does it happen when you are editing any particular files? A screenshot of this one would be particularly helpful i think.


If the answer is no on either of the above, then I would love to take up your offer on setting up a call to debug further please

**Reply by valhuber** [reply]
@dmilburn Hey there...

The rouge printout is not reproducible.  I will send log when I see it, but there's a bunch of obvious gibberish in setting up the port.

I will definitely check for local host next time Browser fails to load.

That said, I would still like to take you up on your offer, but to explore how to better create projects that can be saved to git.  My email is valjhuber@gmail.com.

Regards, and thanks for your diligence on all this...
Val

**Reply by dmilburn** [reply]
thanks for your response and for the offer to keep chatting! unfortunately, my codespaces work keeps me pretty busy so i won't have time to schedule a meeting otherwise. but i really appreciate your diligence on reporting issues!

**Reply by valhuber** [reply]
@dmilburn I can confirm that the port forwarding failure seems to be going after localhost:

<img width="1003" alt="fails-due-to-local" src="https://user-images.githubusercontent.com/5579631/190527400-bb1843b3-b3b1-40c6-aea6-b61236fe4911.png">

**Reply by valhuber** [reply]
@dmilburn Ah!  And here is an example of the "garbled terminal".  Just restarted the server and it was fine...

<img width="1536" alt="garbled-terminal" src="https://user-images.githubusercontent.com/5579631/190532363-ab46283e-eb83-470b-b054-b7fc7185637a.png">

**Reply by dmilburn** [reply]
Wonderful, thanks for confirming and for the screen shots. I've filed an internal issue for an engineer on the port forwarding team to debug further. I can't guarantee when it will be prioritized but it is in our backlog. Thanks again for reporting
